sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,276 @@
+with different `module-id` attribute, the second one is not copied to the
+device image with the first function, but dependency between those device images
+is recorder instead.
+
+After `SYCL_EXTERNAL` functions are arranged into a separate device image(s),
+all non-`SYCL_EXTERNAL` functions are internalized to avoid multiple definition
+errors during runtime linking.
+Device images with `SYCL_EXTERNAL` functions will also have a list of names
+of exported functions.
+
+**NOTE**: If device code split is enabled, it seems reasonable to perform
+exports arrangement before device code split procedure.
+
+In orger to collect information about imported symbols `sycl-post-link` looks
+through LLVM IR and for each declared but not defined symbol records its name,
+except the following cases:
+- Declarations with `__` prefix in demangled name are not recorded as imported
+  functions
+  - Declarations with `__spirv_*` prefix should not be recorded as dependencies
+  since they represent SPIR-V operations and will be transformed to SPIR-V
+  instructions during LLVM->SPIR-V translation.
+- Based on some attributes which could be defined later
+  - This is needed to have possibility to call device-specific builtins not
+    starting with `__` by forward-declaring them in DPC++ code
+
+**NOTE**: If device code split is enabled, imports collection is performed after
+split and it is performed on splitted images.
+
+All collected information is attached to a device image via properties
+mechanism.
+
+Each device image is supplied with an array of property sets:
+```
+struct pi_device_binary_struct {
+...
+  // Array of property sets
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+
+```
+Each property set is represent by the following struct:
+```
+// Named array of properties.
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+```
+It contains name of property set and array of properties. Each property is
+represented by the following struct:
+```
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+List of imported symbols is represented as a single property set with name
+`ImportedSymbols` recorded in the `Name` field of property set.
+Each property in this set holds name of the particular imported symbol recorded
+in the `Name` field of the property.
+List of exported symbols is represented in the same way, except the
+corresponding set has the name `ExportedSymbols`.
+
+### DPC++ runtime changes
+
+DPC++ RT performs *device images collection* task by grouping all device
```

Comment:
```suggestion
DPC++ RT performs *device images dependence resolution* task by grouping all device
```

Reply:
Please note that this is not `symbol resolution` operation, which is performed by dynamic linkers while loading an executable: if the same symbol is available in two libraries, dynamic linker is able to ignore the second definition and use the first one it encountered. However, in our case we do not dictate which exact symbol will be used when it is needed - we can only combine a few device images together and pass them further to low-level runtimes. If we go with resolution, then I suggest that we remove asterisks, because they were intended to emphasize `collection` word

Reply:
Yes, I understand this is not the same on host. As I can see, collection is a part of dependence resolution, which needs to happen anyway. I'd appreciate describing it in more details (already requested in couple of places above). Thanks!